homeassistant/components/cast/media_player.py

```diff
@@ -346,12 +392,25 @@ def async_cast_discovered(discover: ChromecastInfo):
             _LOGGER.debug("Discovered chromecast with same UUID: %s", discover)
             self.hass.async_create_task(self.async_set_cast_info(discover))
 
+        def async_cast_removed(discover: ChromecastInfo):
+            """Handle removal of Chromecast."""
+            if self._cast_info.uuid is None:
+                # We can't handle empty UUIDs
+                return
+            if self._cast_info.uuid != discover.uuid:
+                # Removed is not our device.
+                return
+            _LOGGER.debug("Removed chromecast with same UUID: %s", discover)
+            self.hass.async_create_task(self.async_del_cast_info(discover))
+
         async def async_stop(event):
             """Disconnect socket on Home Assistant stop."""
             await self._async_disconnect()
 
         async_dispatcher_connect(self.hass, SIGNAL_CAST_DISCOVERED,
                                  async_cast_discovered)
+        async_dispatcher_connect(self.hass, SIGNAL_CAST_REMOVED,
```

Comment:
Don't we need to remove the SIGNAL_CAST_DSICOVERED + SIGNAL_CAST_REMOVED subscriptions?

Reply:
Yes, we should do that in `async_will_remove_from_hass`